GH-109975: Copyedit 3.13 What's New: Release Highlights (#122958)
python/cpython: Doc/whatsnew/3.13.rst

.. This section singles out the most important changes in Python 3.13.
   Brevity is key.

> Python 3.13 is the latest stable release of the Python programming language,

Reviewer: It's not quite the latest stable, there's still a month and a half. "will be"?

with a mix of changes to the language, the implementation and the standard library.
The biggest changes include a new `interactive interpreter
<whatsnew313-better-interactive-interpreter_>`_,
experimental support for running in a `free-threaded mode
<whatsnew313-free-threaded-cpython_>`_ (:pep:`703`),
and a `Just-In-Time compiler <whatsnew313-jit-compiler_>`_ (:pep:`744`).

Error messages continue to improve, with tracebacks now highlighted in color
by default. The :func:`locals` builtin now has :ref:`defined semantics
<whatsnew313-locals-semantics>` for changing the returned mapping,
and type parameters now support default values.

The library changes contain removal of deprecated APIs and modules,
as well as the usual improvements in user-friendliness and correctness.
Several legacy standard library modules have now `been removed
<whatsnew313-pep594_>`_ following their deprecation in Python 3.11 (:pep:`594`).

> This article doesn't attempt to provide a complete specification of all new features,

Reviewer: max 80 chars (here and elsewhere) https://devguide.python.org/documentation/markup/#use-of-whitespace

Author: I've wrapped for the edits made here ("Release Highlights") rather than re-wrapping the entire document. Will keep 80 in mind as other edits continue.

but instead gives a convenient overview.

> For full details, you should refer to the documentation,

Reviewer: ```suggestion
For full details refer to the documentation,
```

such as the  :ref:`Library Reference <library-index>`
and :ref:`Language Reference <reference-index>`.

> If you want to understand the complete implementation and design rationale for a change,

Reviewer: ```suggestion
To understand the complete implementation and design rationale for a change,
```

* A greatly improved :ref:`interactive interpreter
  <whatsnew313-better-interactive-interpreter>` and
  :ref:`improved error messages <whatsnew313-improved-error-messages>`.
* :pep:`667`: The :func:`locals` builtin now has
  :ref:`defined semantics <whatsnew313-locals-semantics>` when mutating the
  returned mapping. Python debuggers and similar tools may now more reliably
  update local variables in optimized scopes even during concurrent code
  execution.
* :pep:`703`: CPython 3.13 has experimental support for running with the
  :term:`global interpreter lock` disabled when built with ``--disable-gil``.

>   See :ref:`Free-threaded CPython <whatsnew313-free-threaded-cpython>` for more details.

Reviewer: ... or installed via the official installer on macOS or Windows. Perhaps we don't want to mention this here to keep it brief? Or perhaps leave out "when built with ..." here and have all the details in its own section?

* :pep:`744`: A basic :ref:`JIT compiler <whatsnew313-jit-compiler>` was added.
  It is currently disabled by default (though we may turn it on later).

>   Performance improvements are modest -- we expect to be improving this

Reviewer: ```suggestion
  Performance improvements are modest -- we expect to improve this
```

  over the next few releases.
* Color support in the new :ref:`interactive interpreter
  <whatsnew313-better-interactive-interpreter>`,
  as well as in :ref:`tracebacks <whatsnew313-improved-error-messages>`
  and :ref:`doctest <whatsnew313-doctest>` output.

>   This can be disabled throughhe :envvar:`PYTHON_COLORS` and |NO_COLOR|_

Reviewer: ```suggestion
  This can be disabled through the :envvar:`PYTHON_COLORS` and |NO_COLOR|_
```

  is blocked during :term:`finalization <interpreter shutdown>`.
* The :mod:`argparse` module now supports deprecating command-line options,
  positional arguments, and subcommands.
* The new functions :func:`base64.z85encode` and :func:`base64.z85decode`
  support encoding and decoding `Z85 data <https://rfc.zeromq.org/spec/32/>`_.
* The :mod:`copy` module now has a :func:`copy.replace` function,
  with support for many builtin types and any class defining
  the :func:`~object.__replace__` method.
* The :mod:`dbm.sqlite3` module is now the default :mod:`dbm` backend.
* The :mod:`os` module has a suite of new functions for working with Linux's
  timer notification file descriptors.
* The :mod:`random` module now has a :ref:`command-line interface <random-cli>`.

Security improvements:

* :func:`ssl.create_default_context` sets :data:`ssl.VERIFY_X509_PARTIAL_CHAIN`
  and :data:`ssl.VERIFY_X509_STRICT` as default flags.

C API improvements:

* The :c:data:`Py_mod_gil` slot is now used to indicate that an extension module
  supports running with the :term:`GIL` disabled.
* The :doc:`PyTime C API </c-api/time>` has been added,
  providing access to system clocks.

> * :c:type:`PyMutex` is a new lightweight mutex that occupies a single byte.

Author: @vstinner / @serhiy-storchaka -- is there anything you want to highlight particularly in terms of C-API changes for the *release highlights* section of What's New? These were my suggestions but very happy to change them.

A

Reviewer 2: Nothing particular, but entries in the "C API Changes" section should be reordered. First new functions, then changes in existing functions. Functions with same prefix (e.g. `PyDict_`) or similar purpose should be mentioned next to one other.

You can highlight adding groups of functions with similar effect: for example adding functions that return  strong references rather than borrowed references or adding functions that return different codes for error and the absent item (without raising an exception in the latter case).

New typing features:

* :pep:`696`: Type parameters (:data:`typing.TypeVar`, :data:`typing.ParamSpec`,
  and :data:`typing.TypeVarTuple`) now support defaults.
* :pep:`702`: The new :func:`warnings.deprecated` decorator adds support
  for marking deprecations in the type system.
* :pep:`705`: :data:`typing.ReadOnly` can be used to mark an item of a
  :class:`typing.TypedDict` as read-only for type checkers.
* :pep:`742`: :data:`typing.TypeIs` provides more intuitive
  type narrowing behavior, as an alternative to :data:`typing.TypeGuard`.

Platform support:

* :pep:`730`: Apple's iOS is now an officially supported platform,
  at :pep:`tier 3 <11>`.
  Official Android support (:pep:`738`) is in the works as well.

> * ``wasm32-wasi`` is now a supported as a :pep:`tier 2 <11>` platform.

Reviewer: Link these PEP 11s directly to the tier 3 and 2 headings?

* ``wasm32-emscripten`` is no longer an officially supported platform.

> Important deprecations, removals or restrictions:

Reviewer: ```suggestion
Important removals:
```
These are all removals. 

Or will we add others later? Could update the heading later.

Author: Copied from the 3.12 heading, I've looked through now and I don't see any new deprecation that jumps out as big enough to highlight here -- but can always revisit.

* :ref:`PEP 594 <whatsnew313-pep594>`: The remaining 19 "dead batteries"
  have been removed from the standard library:
  :mod:`!aifc`, :mod:`!audioop`, :mod:`!cgi`, :mod:`!cgitb`, :mod:`!chunk`,
  :mod:`!crypt`, :mod:`!imghdr`, :mod:`!mailcap`, :mod:`!msilib`, :mod:`!nis`,
  :mod:`!nntplib`, :mod:`!ossaudiodev`, :mod:`!pipes`, :mod:`!sndhdr`, :mod:`!spwd`,
  :mod:`!sunau`, :mod:`!telnetlib`, :mod:`!uu` and :mod:`!xdrlib`.
* Remove the :program:`!2to3` tool and :mod:`!lib2to3` module
  (deprecated in Python 3.11).
* Remove the :mod:`!tkinter.tix` module (deprecated in Python 3.6).
* Remove :func:`!locale.resetlocale()`.
* Remove :mod:`!typing.io` and :mod:`!typing.re`.
* Remove chained :class:`classmethod` descriptors.

Release schedule changes:

:pep:`602` ("Annual Release Cycle for Python") has been updated
to extend the full support ('bugfix') period for new releases to two years.
This updated policy means that:

> * Python 3.9---3.12 have one and a half years of full support,

Reviewer: ```suggestion
* Python 3.9--3.12 have one and a half years of full support,
```

[Microsoft](https://learn.microsoft.com/en-us/style-guide/punctuation/dashes-hyphens/enes), [Google](https://developers.google.com/style/hyphens#number-range) and [Apple](https://support.apple.com/en-gb/guide/applestyleguide/apsg7af4f5d0/1.0/web/1.0#apd3abe21dfdd7f4) style guides say to use an en dash here.

* Add :func:`base64.z85encode` and :func:`base64.z85decode` functions which allow encoding
  and decoding Z85 data.

>   See `Z85 specification <https://rfc.zeromq.org/spec/32/>`_ for more information.

Reviewer: ```suggestion
  See the `Z85 specification <https://rfc.zeromq.org/spec/32/>`_ for more information.
```
